nntrainer/include/delegate.h

```diff
@@ -0,0 +1,179 @@
+// SPDX-License-Identifier: Apache-2.0-only
+/**
+ * Copyright (C) 2020 Parichay Kapoor <pk.kapoor@samsung.com>
+ *
+ * @file	delegate.h
+ * @date	7 Aug 2020
+ * @brief	This is Delegate Class for the Neural Network
+ * @see		https://github.com/nnstreamer/nntrainer
+ * @author	Parichay Kapoor <pk.kapoor@samsung.com>
+ * @bug		No known bugs except for NYI items
+ *
+ * @note This class is experimental and subject to major modifications.
+ *
+ */
+
+#ifndef __DELEGATE_H__
+#define __DELEGATE_H__
+
+namespace nntrainer {
+
+/**
+ * @class Backend
+ * @brief Backend to be used for the operations to use.
+ * @note If some operation is not supported on set backend, that operation
+ * implemented with default backend will be used.
+ *
+ */
+class Backend {
+public:
+  /**
+   * @brief Backend type enumeration
+   */
+  enum class BackendType {
+    armcl, /**< Arm Compute Library for backend */
+    blas,  /**< Libopenblas for backend */
+    base   /**< Internally implemented operations for backend */
+  };
+
+  /**
+   * @brief Construct a new Backend object
+   *
+   * @param backend Backend of be used. Defaults to base backend
+   */
+  Backend(BackendType backend = BackendType::base) : backend(backend) {}
+
+  /**
+   * @brief Set the number of threads for the backend if supported
+   *
+   * @param num_threads Number of threads
+   */
+  void setNumThreads(unsigned int num_threads);
+
+private:
+  /**
+   * @brief backend type stored
+   */
+  BackendType backend;
+
+  /**
+   * @brief number of threads stored
+   */
+  unsigned int num_threads;
+};
+
+/**
+ * @class Device
+ * @brief Device to be used for the operations to run.
+ * @note The selected delegate device will restrict the supported backend.
+ *
+ */
+class Device {
+public:
+  /**
+   * @brief Device type enumeration
+   */
+  enum class DeviceType {
+    cpu, /**< CPU as the device */
+    gpu, /**< GPU as the device */
+    npu  /**< NPU as the device */
+  };
+
+  /**
+   * @brief Construct a new Device object
+   *
+   * @param device Device delegate to be used for the operations to run. Default
+   * device is cpu.
+   */
+  Device(DeviceType device = DeviceType::cpu) :
```

Comment:
IMO, it is better to have ctor with full argument with default so that we can instantiate object with full configuration.

```suggestion
  Device(DeviceType device = DeviceType::cpu, float memory_fraction = 1.0, soft_pacement = false, precision_loss = false) :
```

Reply:
yes, updated.
Also updated for backend class as well.